main.go

```diff
@@ -60,7 +72,44 @@ func (sc *SafeConfig) reloadConfig(confFile string) (err error) {
 	return nil
 }
 
-func probeHandler(w http.ResponseWriter, r *http.Request, conf *Config) {
+func probeHandler(w http.ResponseWriter, r *http.Request, c *Config) {
+
+	if c == nil {
+		sc.RLock()
+		c = sc.C
+		sc.RUnlock()
+	}
+
+	moduleName := r.URL.Query().Get("module")
+	if moduleName == "" {
+		moduleName = "http_2xx"
+	}
+	module, ok := c.Modules[moduleName]
+	if !ok {
+		http.Error(w, fmt.Sprintf("Unknown module %q", moduleName), 400)
+	}
+
+	// If a timeout is configured via the Prometheus header, add it to the request.
+	var prometheusTimeout string
+	if r.Header["X-Prometheus-Scrape-Timeout-Seconds"] != nil {
+		prometheusTimeout = r.Header["X-Prometheus-Scrape-Timeout-Seconds"][0]
+	}
+
+	timeoutSeconds, err := strconv.ParseFloat(prometheusTimeout, 64)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Failed to parse timeout from Prometheus header: %s", err), http.StatusInternalServerError)
+	}
+	if timeoutSeconds == 0 {
+		timeoutSeconds = 10
+	}
+
+	if module.Timeout.Seconds() < timeoutSeconds && module.Timeout.Seconds() > 0 {
+		timeoutSeconds = module.Timeout.Seconds()
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeoutSeconds-*timeoutOffset)*time.Second)
```

Comment:
You're using precision here as time.Duration is an int64. Multiplying it by 1e9 before casting would keep the precision.